internal/pkg/skuba/addons/dex.go

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2019 SUSE LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package addons
+
+import (
+	"github.com/SUSE/skuba/internal/pkg/skuba/dex"
+	"github.com/SUSE/skuba/internal/pkg/skuba/kubernetes"
+	"github.com/SUSE/skuba/internal/pkg/skuba/skuba"
+	skubaconstants "github.com/SUSE/skuba/pkg/skuba"
+	"github.com/pkg/errors"
+	"k8s.io/kubernetes/cmd/kubeadm/app/images"
+)
+
+var gangwayClientSecret string
+
+func init() {
+	registerAddon(kubernetes.Dex, renderDexTemplate, DexCallbacks{})
+}
+
+func (addonRenderContext addonRenderContext) DexImage() string {
+	return images.GetGenericImage(skubaconstants.ImageRepository, "dex",
+		kubernetes.AddonVersionForClusterVersion(kubernetes.Dex, addonRenderContext.Config.ClusterVersion).Version)
+}
+
+func (addonRenderContext addonRenderContext) GangwayClientSecret() string {
+	if len(gangwayClientSecret) == 0 {
+		gangwayClientSecret = dex.GenerateClientSecret()
+	}
+	return gangwayClientSecret
+}
+
+func renderDexTemplate(addonConfiguration AddonConfiguration) string {
+	return dexManifest
+}
+
+type DexCallbacks struct{}
+
+func (DexCallbacks) BeforeApply(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration) error {
+	client, err := kubernetes.GetAdminClientSet()
+	if err != nil {
+		return errors.Wrap(err, "could not get admin client set")
+	}
+	dexCertExists, err := dex.DexCertExists(client)
+	if err != nil {
+		return errors.Wrap(err, "unable to determine if dex certificate exists")
+	}
+	err = dex.CreateCert(client, skubaconstants.PkiDir(), skubaconstants.KubeadmInitConfFile())
+	if err != nil {
+		return errors.Wrap(err, "unable to create dex certificate")
+	}
+	if dexCertExists {
+		if err := dex.RestartPods(client); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (DexCallbacks) AfterApply(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration) error {
+	return nil
+}
+
+const (
```

Comment:
Wondering if we should also add CaaSP metadata into the different objects :thinking: Same for the gangway addon.

Reply:
I added the caasp metadata on the places necessary for the pods to automatically restart if something changes. The real information is saved into the `skuba-config` ConfigMap.